Add Display Support

brentru opened the pull request:
This pull request introduces a new protocol buffer definition for display management in WipperSnapper API v1, integrating display support into the existing messaging system. 

### Coverage
- Implemented new `display.proto` API
- Modified `signal.proto` to include messages from `display.proto` in `MsgRequest`/`MsgResponse` format.
- Added `display.md` API documentation

### Implementation Details:
cc @lorennorman 
I implemented with YAGNI in mind so the changes are only for EPD. However, the API is flexible enough to extend to TFTs and (existing) OLED, LED matrix and LCD character displays.

This will require two extra MQTT topics for implementing the `DisplayRequest` and `DisplayResponse` messages:
- `{USER}/wprsnpr/io-wipper-{UID}/signals/broker/display` - `DisplayRequest` messages should be transmitted over this topic.
- 
- `{USER}/wprsnpr/io-wipper-{UID}/signals/device/display` - `DisplayResponse` messages should be transmitted over this topic
- The file `display.md` provides sequence diagrams for the Request and Response messages.

brentru committed:
Rough draft!

brentru committed:
Add EPD type and panel config

brentru committed:
Fix import path

brentru committed:
rm extra pins

brentru committed:
MagTag is 2.90 not 2.13

brentru committed:
Rm unused panel configuration

brentru committed:
Increase name and message strlens

brentru committed:
Rm width, height. Add name to Response messages

brentru committed:
Add display.md document

lorennorman commented on proto/wippersnapper/display/v1/display.md:
```diff
@@ -0,0 +1,75 @@
+
+ws_display controller->>ws_display hardware: Delete hardware instance
+
+ws_display hardware->>ws_display driver: Delete driver instance
```
missing the return route for `DisplayRemoved` here?

brentru replied:
Yup, I should implement that. Will do on the next commit.

lorennorman commented on proto/wippersnapper/display/v1/display.proto:
```diff
@@ -0,0 +1,121 @@
+  DisplayType type = 1; /** The type of display */
+  string name = 2[(nanopb).max_size = 64]; /** Identifier for the display */
+  oneof interface_type {
+    EpdSpiConfig spi_epd = 3; /** SPI configuration for EPD */
+  }
+  oneof config {
+    EPDConfig epd_config = 4; /** Configuration for EPD display */
+  }
```
I'm guessing these `oneof`s are the spots where this can get expanded for new display types?

brentru replied:
Correct yes, it'd be like `EpdSpiConfig` (epd needs an extra sram pin), `TftSpiConfig`, `i2c.i2cConfig`.. etc

lorennorman commented on proto/wippersnapper/display/v1/display.proto:
```diff
@@ -0,0 +1,121 @@
+ * EPDWriteOptions contains options for writing to EPD displays.
+ */
+message EPDWriteOptions { 
+  int32 text_size = 1; /** Size of the text to write */
```
maybe a little more info in the comment here, is this a size in pixels? does it have upper/lower bounds?

brentru replied:
Sure, will add in next commit.

brentru replied:
Clarified that this is a scale factor for text rendering (1 = 6x8px, 2 = 12x16px, etc.)

lorennorman reviewed (approved):
🤩 looks good! 💯 
mermaid docs are so great to help me (us) out later, and the API itself looks like it goes straight to the point with appropriate affordances for potential later extension.

brentru committed:
Fix for @lorennorman

brentru committed:
Add Tft constraints

brentru committed:
Add TFT displayType

brentru committed:
Bring i2c_output displays into display.proto, refactor!

brentru committed:
Remove i2c output from i2c.proto

brentru committed:
Update signal to rm i2c output

brentru committed:
Remove SSD1306-specific language

brentru committed:
Remove two enums

brentru committed:
Revert all changes after 48fc8bb

brentru committed:
Revert all changes after b7e6d06

lorennorman commented on proto/wippersnapper/display/v1/display.proto:
```diff
@@ -0,0 +1,144 @@
+
+package wippersnapper.display.v1;
+import "nanopb/nanopb.proto";
+import "wippersnapper/i2c/v1/i2c.proto";
```
is this import still needed? (protoc is throwing a warning that it is unused)

brentru replied:
@lorennorman  Unused, you can remove it.

lorennorman committed:
remove unused import

lorennorman commented on proto/wippersnapper/display/v1/display.proto:
```diff
@@ -0,0 +1,144 @@
+/**
+ * DisplayAddedorReplaced confirms a display was added or replaced.
+ */
+message DisplayAddedorReplaced {
```
inconsistent capitalization, should be `DisplayAddedOrReplaced`

brentru replied:
Missed that, thanks. Addressed in the commit below.

brentru committed:
Change DisplayAddedOrReplaced

brentru committed:
Refactor per Loren review today

brentru committed:
Update for Loren

lorennorman commented on proto/wippersnapper/display/v1/display.proto:
```diff
@@ -0,0 +1,151 @@
+
+/**
+ * DisplayAddOrReplace adds a new display or replaces an existing one.
+ */
+message DisplayAddOrReplace {
+  DisplayType type       = 1; /** The type of display */
+  DisplayDriver driver   = 2; /** The specific driver for the display */
+  oneof interface_type {
+    EpdSpiConfig spi_epd = 3; /** SPI configuration for EPD. */
+    TftSpiConfig spi_tft = 4; /** SPI configuration for TFT. */
+  }
+  oneof config {
+    EPDConfig config_epd = 5; /** Configuration for EPD. */
+    TftConfig config_tft = 6; /** Configuration for TFT. */
+  }
+}
```
I think we still need `name` here for other reasons, right? (I assume you're using it as an identifier to look up later for Writes, Replaces, and Removes.)

brentru replied:
Oh, yeah, it shouldn't have been fully removed...

brentru committed:
Put back name field

lorennorman commented on proto/wippersnapper/display/v1/display.proto:
```diff
@@ -0,0 +1,154 @@
+  EPDWriteOptions options = 2; /** Options for writing to the EPD */
+  int32 width             = 3;   /** Width of the EPD display, in pixels */
+  int32 height            = 4;   /** Height of the EPD display, in pixels */
+}
```
we can collapse `text_size` into the `EPDConfig` message, now

brentru replied:
Addressed in 1d49c0

brentru committed:
Collapse EPDWriteOptions into EPDConfig, add text_size to tftConfig

brentru commented:
@lorennorman  I have removed `EPDWriteOptions`. I moved `text_size` into `EPDConfig` and `TFTConfig`.

The form for TFTs does not currently show an option for `text_size`, but it should be able to expose that now.
